### PR TITLE
OWASP ZAP Dot Net API related changes

### PR DESCRIPTION
**Pull request by gmaran23**
## OWASP ZAP Dot Net API related changes

-----------------------------------------------------------------------------------------------------
### + 	newly added file
-----------------------------------------------------------------------------------------------------
1. **src/org/zaproxy/zap/extension/api/DotNetAPIGenerator.java**
		(new file to auto generate CSharp class files for ZAP's components)

-----------------------------------------------------------------------------------------------------
#### Note:
-----------------------------------------------------------------------------------------------------
1. Binary (**OWASPZAPDotNetAPI.dll**) available at [https://www.nuget.org/packages/OWASPZAPDotNetAPI]

### CODE REVIEW

**Comment by gmaran23** [comment]
Guys, Is there some change that needs to be done for this pull request to be merged?

**Comment by psiinon** [comment]
Sorry, we should have got back to you on this :(
Its all looking good, but we decided that it might be better if the APIs were split into new projects, eg zap-api-java, zap-api-dotnet etc
This should make it easier to structure the projects a bit more sensibly, and assign different permissions as well - so you could get write permissions to the dotnet one :)
The plan was to move the java one first, but thats still in progress :/
But we could just create the zap-api-dotnet one and then you could submit a PR to that?

**Comment by gmaran23** [comment]
Simon, Makes sense. Let me know once you have created a project named zap-api-dotnet.  
I am assuming that the files related to the dotnet api could go in to the zap-api-dotnet folder, however the *DotNetApiGenrator.java* file would still reside in *src/org/zaproxy/zap/extension/api/* correct?

Looking forward to create a PR for the zap-api-dotnet project.

**Comment by psiinon** [comment]
Just created https://github.com/zaproxy/zap-api-dotnet and if you accept the request you should have write access to it.
Please set it up whatever way makes sense for dot NET projects, and include instructions for building the API.
Many thanks!

**Comment by psiinon** [comment]
Oh, and yes, the DotNetApiGenrator.java file should still reside in src/org/zaproxy/zap/extension/api/ :)

File: src/org/zaproxy/zap/extension/api/DotNetAPIGenerator.java [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+			this.generateCSharpElement(other, imp.getPrefix(), "other", out);
+		}
+		out.write("\t}\n");
+		out.write("}\n");
+		out.close();
+	}
+
+	public static void main(String[] args) throws Exception {
+		// Command for generating a DotNet version of the ZAP API		
+		DotNetAPIGenerator dnapi = new DotNetAPIGenerator("../zap-api-dotnet/src/OWASPZAPDotNetAPI/OWASPZAPDotNetAPI/Generated", true);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/DotNetAPIGenerator.java]: The `optional` parameter should be `false`, core APIs are always available.

**Comment by thc202** [comment]
Could you fixup the commits?

**Comment by gmaran23** [comment]
thc202, nice catch. Fixed it. Let me know if anything else needs amending.

**Comment by thc202** [comment]
Could you fixup the commits?

**Comment by thc202** [comment]
Oh, and the commit message needs to be updated accordingly.

**Comment by gmaran23** [comment]
Sorry. How imbecile. I have updated the commit message. I have changed the optional parameter to false. Is there something else that needs fixing?

**Comment by thc202** [comment]
Not anything in the code, just fixup the commits of the pull request (e.g. to remove API changes that were already added to other repo).

**Comment by kingthorin** [comment]
@gmaran23 is a fixup of this coming?

**Comment by gmaran23** [comment]
I am confused on what commits to fix up. Would it make sense if I create a new commit (updating the Dot Net API with the most recent version of ZAP) and forget about #1719 pull request? Meaning... discard the pull request #1719, and then create a new pull request including just relevant commits? 
      

    On Sunday, 21 February 2016 5:13 AM, kingthorin <notifications@github.com> wrote:
 
 

 @gmaran23 is a fixup of this coming?—

**Comment by thc202** [comment]
You can fixup all of the commits  except the first (IMHO).
For example:
$ git rebase -i 2015-06-25
Then set all except `7cbffcc` to "fixup", this would leave your branch with just the addition of the file DotNetAPIGenerator.java with the latest changes but without all the other unnecessary commits.

You can then rebase on upstream/develop to bring the branch up-to-date and then amend the commit to add the generator to Ant task "generate-apis" (this change was lost/reverted in one of the commits).

What do you think?

**Comment by kingthorin** [comment]
@gmaran23  any feedback?

**Comment by gmaran23** [comment]
I prefer closing this pull request and starting over with a new one. I am not a Git expert and to me it seems like the easiest way to go.

Will submit a new PR in a few weeks.

**Comment by kingthorin** [comment]
Thanks for getting back to us. Will watch for the new one.

**Comment by psiinon** [comment]
We are planning on splitting the APIs out into separate repos. @thc202 any progress on this?

**Comment by thc202** [comment]
The .NET API is already in its own [repo](https://github.com/zaproxy/zap-api-dotnet) and IIRC the generators would still be added to zaproxy repo. So, a new pull request with the generator is what's left for this API.

**Comment by binarymist** [comment]
PR submitted: https://github.com/zaproxy/zaproxy/pull/2625

**Pull request closed without merging**